### ISSUE TITLE
Break up single/double quote rules

### ISSUE THREAD

**Issue by corysimmons** [stylelint/stylelint#704]
I want

```css
[class="foo"] {
  content: '';
}
```

**Comment by davidtheclark** [comment]
See https://github.com/stylelint/stylelint/issues/96

The main point there:

> I am not sure that distinguishing selector attribute quotes from all other "string" quotes makes a ton of sense unless we get even more granular: if the principle is that we're going to allow the user to specify a quote style in every location in which a string can be used, then it would be consistent; if not, isn't it a kind of arbitrary split to single out selector attributes?

I don't want to try to splitting out quote rules unless someone can come up with a sane schema for doing so. Ideas?

**Comment by corysimmons** [comment]
I don't think it's arbitrary to have double quotes for selectors and single quotes for everything else.

I see stuff like this all the time:

```css
[class="foo"] {
  conten: '';
  background: url('foo.jpg');
}
```

**Comment by corysimmons** [comment]
Maybe we should just take this convo over to that issue instead?

**Comment by davidtheclark** [comment]
Let's keep the conversation here because that issue is old. The rule is already built: this is just an additional feature request.

I myself am not interested in building in the complexity of expecting different quotation marks in different places. @jeddy3 ?

**Comment by corysimmons** [comment]
Eh I think it'd be cool, but again, not a big deal.

**Comment by jeddy3** [comment]
> I myself am not interested in building in the complexity 

Ditto.

**Comment by davidtheclark** [comment]
The way that this rule technically works makes that it impossible to distinguish the node in which the quotation mark occurs. We check the entire CSS string for quotation marks; we are not checking individual nodes one at a time. So when we run across a quotation mark we do not really know whether it's part of an attribute selector or part of a `content` value: we just know that it's a quotation mark indicating the edge of a string so should match a convention. 

When I wrote the rule, it seemed to me less viable, more possibly-buggy, to try to identify all the possible nodes that *might* contain strings and then check those individually — but if we did do that, it would be easy to add `ignore` or `except` options based on node type. 

If anybody feels that they can confidently *list out all of the CSS nodes that might contain strings*, we could reconsider the workings of this rule. Otherwise, though, I'm afraid this feature request isn't feasible.

**Comment by corysimmons** [comment]
stylelint is built on top of PostCSS right? walk decl; walk selectors.

**Comment by davidtheclark** [comment]
In order to consider being more granular about where the particular quotation mark is we'd need a definitive list of where CSS strings *can* occur, &if that list is finite*. If it's not finite, then this can't really be done as far as I can tell.

I'll leave this issue open in case anybody wants to have a go at that list.

**Comment by jeddy3** [comment]
(issue triage as there will likely be plenty inbound issues after the big release of 4.5)

Closing this due to inactivity. If the request come up again I suggest we point them at this issue in the hope they create that definitive list.

**Comment by kaelig** [comment]
I'd really like this too. For our Sass and our strings we use single quotes (easier to type), however in our attribute selectors `[foo="bar"]`, typing a double quote is worth the trouble because it makes searching the codebase easier (we catch both the markup and the CSS in one search query).

@jeddy3 any way this could become a thing? It seems to be an issue that keeps coming back so I'd love to see this being given another chance by the awesome stylelint team.

**Comment by jeddy3** [comment]
> any way this could become a thing?

@kaelig Thanks for your input. As David [pointed out above](https://github.com/stylelint/stylelint/issues/704#issuecomment-181872385), such an approach, which covers this particular use-case, might be less viable and more bug-prone than what we have now. However, around the time of this issue (a little under a year ago) we only had a fledgeling system of extensibility. Now we have a fully-realised one.  The [plugin system](https://stylelint.io/developer-guide/plugins/) is for precisely these types of scenarios i.e. experimentation. Feel free to experiment with a collection of plugin rules that offer finer-grained control of quotation marks (paying particular attention to those that appear within non-standard constructs). If the approach turns out to be viable and not error-prone then we can consider migrating it into core and replacing the existing rules.  Does that sound good?

I'll relabel this issue as "plugin" for now.

**Comment by kaelig** [comment]
Thanks @jeddy3 - I'll look into creating a plugin, I feel like it could be useful to the community.
